bigchaindb/core.py

```diff
@@ -444,15 +444,20 @@ def has_previous_vote(self, block):
                 but the vote is invalid.
 
         """
-        if block['votes']:
-            for vote in block['votes']:
-                if vote['node_pubkey'] == self.me:
-                    if not util.verify_vote_signature(block, vote):
-                        raise exceptions.ImproperVoteError(
-                            'Block {} already has an incorrectly signed vote from public key {}'
-                        ).format(block['id'], self.me)
-                    return True
-        return False
+        votes = list(r.table('votes').get_all([block['id'], self.me], index='block_and_voter').run(self.conn))
+
+        if len(votes) > 1:
+            raise exceptions.MultipleVotesError('Block {block_id} has {n_votes} votes from public key {me}'
+                                     .format(block_id=block['id'], n_votes=str(len(votes)), me=self.me))
+        has_previous_vote = False
+        for vote in votes:
```

Comment:
I still don't understand this for loop. You already raise an exception before `if len(votes) > 1` and at this point you either have 0 or 1 vote